bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/infoproviders/AddonsInfoProvider.java

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (c) 2010-2023 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.addon.infoproviders;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.regex.PatternSyntaxException;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.OpenHAB;
+import org.openhab.core.addon.AddonDiscoveryMethod;
+import org.openhab.core.addon.AddonInfo;
+import org.openhab.core.addon.AddonInfoListReader;
+import org.openhab.core.addon.AddonInfoProvider;
+import org.openhab.core.addon.AddonMatchProperty;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.thoughtworks.xstream.XStreamException;
+import com.thoughtworks.xstream.converters.ConversionException;
+
+/**
+ * The {@link AddonsInfoProvider} reads all {@code userdata/addons/*.xml} files, each of which
+ * should contain a list of addon.xml elements, and convert their combined contents into a list
+ * of {@link AddonInfo} objects can be accessed via the {@link AddonInfoProvider} interface.
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = AddonInfoProvider.class, name = AddonsInfoProvider.SERVICE_NAME)
+public class AddonsInfoProvider implements AddonInfoProvider {
+
+    public static final String SERVICE_NAME = "addons-info-provider";
+
+    private static final boolean TEST_ADDON_DEVELOPER_REGEX_SYNTAX = true;
+
+    private final Logger logger = LoggerFactory.getLogger(AddonsInfoProvider.class);
+    private final String folder = OpenHAB.getUserDataFolder() + File.separator + "addons";
+    private final Set<AddonInfo> addonInfos = new HashSet<>();
+
+    @Activate
+    public AddonsInfoProvider() {
+        initialize();
+        if (TEST_ADDON_DEVELOPER_REGEX_SYNTAX) {
```

Comment:
Can this every be anything different than `true`...? If it was just for dev purposes, I'd rather remove it now.

Reply:
I extended the .xsd schema for the addon.xml so when a developer creates a new binding the SAT tests will validate the xml against the xsd. HOWEVER the xsd validation is not sufficient to test if the matchProperty regex is good or not. This risks that an addon may cause an exception in OH core due to a bad regex. So I added this to test the regexes. I imagine one would run those tests from time to time on the whole OH build, but not always..

Reply:
The difficulty is that if OH would throw a bad regex exception at runtime, it would not know which addon had caused it (since the error comes from parsing an xml file)..

Reply:
As you suggested in an earlier comment, we could do this on a thread anyway.

Reply:
I will do it always on start-up. We only have about 70 addon xml entries to load, and about 80 regexes to check, so it should not take that long.

Reply:
ok, let's hope that it does not have any negative impact on startup then.

Reply:
@kaikreuzer in my full build tests I did not notice any delay